File: 747.Largest-Number-At-Least-Twice-of-Others.go
// https://leetcode.com/problems/split-linked-list-in-parts/description/
//
// algorithms
// Easy (39.88%)
// Total Accepted:    34.5k
// Total Submissions: 86.5k
// beats 100.0% of golang submissions

package leetcode

func dominantIndex(nums []int) int {
	length := len(nums)

	if length == 1 {
		return 0
	}

	first, second, res := -1, -1, -1
	for idx, num := range nums {
		if num > first {
			second = first
			first = num
			res = idx
		} else if num > second {
			second = num
		}
	}

	if first >= 2*second {
		return res
	}
	return -1
}

File: 747.Largest-Number-At-Least-Twice-of-Others.py
# https://leetcode.com/problems/largest-number-at-least-twice-of-others/
#
# algorithms
# Easy (40.8%)
# Total Accepted:    59,586
# Total Submissions: 146,035


class Solution(object):
    def dominantIndex(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        st, nd = float('-inf'), float('-inf')
        st_idx = -1

        for idx, n in enumerate(nums):
            if n > st:
                nd = st
                st, st_idx = n, idx

            elif n > nd:
                nd = n

        if st >= 2 * nd:
            return st_idx

        return -1
